data/src/main/java/com/linkedin/data/Data.java

```diff
@@ -491,11 +488,49 @@ private static void traverse(Object obj, TraverseCallback callback, CycleChecker
             cycleChecker.startMap(map);
             callback.startMap(map);
             Iterable<Map.Entry<String, Object>> orderedEntrySet = callback.orderMap(map);
-            for (Map.Entry<String, Object> entry : orderedEntrySet)
+
+            //
+            // If the ordered entry set is null, use Java 8 forEach to avoid intermediary object
+            // creation for better performance.
+            //
+            if (orderedEntrySet == null)
```

Comment:
1
Previously it didn't handle the `orderedEntrySet` is null case, is this condition handling here a bug fix?
2
And line 490's `callback.orderMap(map);` is getting the map's entrySet internally and return orderedEntrySet.
But here I see we are using `map.forEach`... is it actually going to do anything in the case that there is no entries?

Reply:
1. Not a bug fix. I changed the contract to return null if we were ok with the map's natural ordering. This enables faster iteration on the map and lesser garbage since we can directly use forEach. It is backward compatible as well.

2. orderMap will return an ordered iterable if it cares about order, else null. We will process it accordingly.

Reply:
I see, so it depends on what actual map it is. 
If it is a tree map, then the orderedEntrySet will not be null 
in hash map, the orderedEntrySet will be null and goes into this condition.
In both cases the entries are there.
Good learn for me